### PR TITLE
Malte lig 1897 implement run worker label matching

### PR DESCRIPTION
**Pull request by MalteEbner**
## What has changed?
- generated from the new spec:
  - `get_docker_runs_scheduled_by_state_and_labels` with `labels`
  - `DockerRunScheduledCreateRequest` with `runs_on`
  - `DockerRunScheduledData` with `runs_on`
  - `DockerRunScheduledUpdateRequest`with `runs_on`
- `ApiWorkflowClient.schedule_compute_worker_run`: added `runs_on` argument
- `ApiWorkflowClient.register_compute_worker`: added `labels` argument
- mock the `urllib3.request.urlencode` to use `doseq=True`. Otherwise using arrays in the query fails. This is needed for `get_docker_runs_scheduled_by_state_and_labels`

## How was it tested? + corresponding lightly-core PR
see https://github.com/lightly-ai/lightly-core/pull/2037

### CODE REVIEW

File: lightly/api/__init__.py [hunk 1]
```diff
@@ -5,3 +5,10 @@
 
 from lightly.api.api_workflow_client import ApiWorkflowClient
 from lightly.api.api_workflow_compute_worker import ArtifactNotExist
+
+# Mock the urlencode to use doseq, so that arrays in the query are properly encoded
+from six.moves.urllib.parse import urlencode
```
Comment (japrescott) [inline lightly/api/__init__.py:10]: will this not be overwritten when we generate the code again?

File: tests/api_workflow/test_api_workflow_compute_worker.py [hunk 2]
```diff
@@ -368,6 +370,7 @@ def test_get_compute_worker_state_and_message_OPEN() -> None:
         state=DockerRunScheduledState.OPEN,
         created_at=0,
         last_modified_at=1,
+        runs_on=[6]
```
Comment (japrescott) [inline tests/api_workflow/test_api_workflow_compute_worker.py:373]: I am going to put on my Malte hat and ask you to write tests :)

File: lightly/api/__init__.py (hunk 1 above)
Comment (MalteEbner) [reply lightly/api/__init__.py:10]: This is not part of the generated code. The generated code is only under `lightly/openapi_generated`

File: lightly/api/__init__.py [hunk 3]
```diff
@@ -5,3 +5,10 @@
...
+def doseq_urlencode(fields):
+    return urlencode(fields, doseq=True)
+import urllib3.request
+urllib3.request.urlencode = doseq_urlencode
```
Comment (guarin) [inline lightly/api/__init__.py:14]: This is pretty hacky as it changes urllib3 under the hood. A less intrusive alternative is to patch the requests method on the rest client:
```
from lightly.openapi_generated.swagger_client.rest import RESTClientObject

def patch_rest_client(rest_client):
    request = rest_client.request

    def request_patched(self, method, url, query_params=None, headers=None,
                    body=None, post_params=None, _preload_content=True,
                    _request_timeout=None):
        if query_params is not None:
            new_query_params = []
            for name, value in query_params:
                if isinstance(value, list):
                    new_query_params.extend([(name, val) for val in value])
                else:
                    new_query_params.append((name, value))
            query_params = new_query_params
        return request(self, method=method, url=url, query_params=query_params, headers=headers, body=body, post_params=post_params, _preload_content=_preload_content, _request_timeout=_request_timeout)

    rest_client.request = request_patched

patch_rest_client(RESTClientObject)
```

What it does is convert query parameters that look like this: `[('labels', ['A', 'B', 'C'])]` into a list of simple parameters: `[('labels', 'A'), ('labels', 'B'), ('labels', 'C')]` which is supported by swagger codegen.

File: tests/api_workflow/test_api_workflow_compute_worker.py (hunk 2 above)
Comment (MalteEbner) [reply tests/api_workflow/test_api_workflow_compute_worker.py:393]: Good point, I added them in the commit `added tests for schedule_compute_worker_run`

**Review by guarin — COMMENTED** [review]
Looks already pretty good. Added some comments :)

File: lightly/api/__init__.py (hunk 3 above)
Comment (MalteEbner) [reply lightly/api/__init__.py:14]: Thanks, good idea!
However, this still messes up with a library, but luckily only a generated one.

File: lightly/api/api_workflow_compute_worker.py [hunk 4]
```diff
@@ -143,6 +147,7 @@ def schedule_compute_worker_run(
         lightly_config: Optional[Dict[str, Any]] = None,
         selection_config: Optional[Union[Dict[str, Any], SelectionConfig]] = None,
         priority: str = DockerRunScheduledPriority.MID,
+        runs_on=None
```
Comment (guarin) [inline lightly/api/api_workflow_compute_worker.py:150]: Type missing.

File: lightly/api/patch_rest_client.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+from typing import Type
+
+
+def patch_rest_client(rest_client: Type):
```
Comment (guarin) [inline lightly/api/patch_rest_client.py:4]: Please add a short description what this does and why we need it.

File: lightly/api/api_workflow_compute_worker.py [hunk 6]
```diff
@@ -68,19 +68,23 @@ def ended_successfully(self) -> bool:
 
 
 class _ComputeWorkerMixin:
-    def register_compute_worker(self, name: str = "Default") -> str:
+    def register_compute_worker(self, name: str = "Default", labels: List[str] = None) -> str:
```
Comment (guarin) [inline lightly/api/api_workflow_compute_worker.py:71]: Type should be `Optional[List[str]]`

**Review by guarin — COMMENTED** [review]
Great! Just some minor things missing.

**Review by guarin — APPROVED** [review]
💯

**Pull request merged**